File: perl5/sync_status.pl
#!/usr/bin/perl
# sync_status.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:skills/sync-utils/scripts/sync_status.py
# auch in: OpenClaw@gateway2:skills/sync-utils/scripts/sync_status.py
# Erzeugt: 2026-08-20 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use JSON;
use File::Find;
use File::Spec;
use POSIX qw(strftime);

# Sync Status - Zeigt Status aller Skills

my $CLAWHUB_DIR = "/home/openclaw/.openclaw/workspace/skills";
my $GIT_DIR = "/home/openclaw/.openclaw/workspace/git/skills";
my $STATE_FILE = "/home/openclaw/.openclaw/workspace/db/sync_state.json";

sub get_max_mtime {
    my ($path) = @_;
    my $max_mtime = 0;
    
    find(sub {
        return if -d $_;
        return if $_ eq '.' || $_ eq '..';
        return if $File::Find::dir =~ /\.git/;
        my $full_path = $File::Find::name;
        my @stat = stat($full_path);
        if (@stat && $stat[9] > $max_mtime) {
            $max_mtime = $stat[9];
        }
    }, $path);
    
    return $max_mtime;
}

sub check_skill_status {
    my ($skill_name) = @_;
    
    my $clawhub_path = File::Spec->catdir($CLAWHUB_DIR, $skill_name);
    my $git_path = File::Spec->catdir($GIT_DIR, $skill_name);
    
    my %status = (
        name => $skill_name,
        in_clawhub => -d $clawhub_path,
        in_git => -d $git_path,
        has_git_repo => (-d $git_path && -d File::Spec->catdir($git_path, ".git")),
        status => "unknown",
        last_modified => {}
    );
    
    # Status bestimmen
    if ($status{in_clawhub} && !$status{in_git}) {
        $status{status} = "only_clawhub";
    } elsif ($status{in_git} && !$status{in_clawhub}) {
        $status{status} = "only_git";
    } elsif ($status{in_clawhub} && $status{in_git}) {
        # Timestamps vergleichen
        eval {
            my $clawhub_mtime = get_max_mtime($clawhub_path);
            my $git_mtime = get_max_mtime($git_path);
            
            $status{last_modified}->{clawhub} = strftime('%Y-%m-%d %H:%M:%S', localtime($clawhub_mtime));
            $status{last_modified}->{git} = strftime('%Y-%m-%d %H:%M:%S', localtime($git_mtime));
            
            if (abs($clawhub_mtime - $git_mtime) < 60) {
                $status{status} = "synced";
            } elsif ($clawhub_mtime > $git_mtime) {
                $status{status} = "clawhub_newer";
            } else {
                $status{status} = "git_newer";
            }
        };
        if ($@) {
            $status{status} = "error";
        }
    }
    
    return \%status;
}

sub main {
    print "=" x 80 . "\n";
    print "ClawHub ↔ Git Sync Status\n";
    print "=" x 80 . "\n";
    print "Zeitpunkt: " . strftime('%Y-%m-%d %H:%M:%S', localtime()) . "\n\n";
    
    # Alle Skills finden
    my %all_skills;
    
    if (-d $CLAWHUB_DIR) {
        opendir(my $dh, $CLAWHUB_DIR) or die "Cannot open directory $CLAWHUB_DIR: $!";
        while (readdir $dh) {
            next if /^\./;
            my $path = File::Spec->catdir($CLAWHUB_DIR, $_);
            $all_skills{$_} = 1 if -d $path;
        }
        closedir $dh;
    }
    
    if (-d $GIT_DIR) {
        opendir(my $dh, $GIT_DIR) or die "Cannot open directory $GIT_DIR: $!";
        while (readdir $dh) {
            next if /^\./;
            my $path = File::Spec->catdir($GIT_DIR, $_);
            $all_skills{$_} = 1 if -d $path;
        }
        closedir $dh;
    }
    
    # Status-Kategorien
    my %categories = (
        synced => [],
        clawhub_newer => [],
        git_newer => [],
        only_clawhub => [],
        only_git => [],
        error => []
    );
    
    # Status für jeden Skill prüfen
    for my $skill (sort keys %all_skills) {
        my $status = check_skill_status($skill);
        push @{$categories{$status->{status}}}, $status;
    }
    
    # Ausgabe
    my $total_skills = scalar(keys %all_skills);
    print "📊 Gesamt: $total_skills Skills\n\n";
    
    # Synchronisiert
    if (@{$categories{synced}}) {
        print "✅ Synchronisiert (" . scalar(@{$categories{synced}}) . ")\n";
        for my $s (@{$categories{synced}}) {
            print "   - " . $s->{name} . "\n";
        }
        print "\n";
    }
    
    # ClawHub neuer
    if (@{$categories{clawhub_newer}}) {
        print "🔄 ClawHub neuer (" . scalar(@{$categories{clawhub_newer}}) . ")\n";
        for my $s (@{$categories{clawhub_newer}}) {
            print "   - " . $s->{name} . " (ClawHub: " . $s->{last_modified}->{clawhub} . ")\n";
        }
        print "\n";
    }
    
    # Git neuer
    if (@{$categories{git_newer}}) {
        print "🔄 Git neuer (" . scalar(@{$categories{git_newer}}) . ")\n";
        for my $s (@{$categories{git_newer}}) {
            print "   - " . $s->{name} . " (Git: " . $s->{last_modified}->{git} . ")\n";
        }
        print "\n";
    }
    
    # Nur in ClawHub
    if (@{$categories{only_clawhub}}) {
        print "📦 Nur in ClawHub (" . scalar(@{$categories{only_clawhub}}) . ")\n";
        for my $s (@{$categories{only_clawhub}}) {
            print "   - " . $s->{name} . "\n";
        }
        print "\n";
    }
    
    # Nur in Git
    if (@{$categories{only_git}}) {
        print "📁 Nur in Git (" . scalar(@{$categories{only_git}}) . ")\n";
        for my $s (@{$categories{only_git}}) {
            print "   - " . $s->{name} . "\n";
        }
        print "\n";
    }
    
    # Fehler
    if (@{$categories{error}}) {
        print "❌ Fehler (" . scalar(@{$categories{error}}) . ")\n";
        for my $s (@{$categories{error}}) {
            print "   - " . $s->{name} . "\n";
        }
        print "\n";
    }
    
    # State-File Info
    if (-f $STATE_FILE) {
        open(my $fh, '<', $STATE_FILE) or die "Cannot open file $STATE_FILE: $!";
        my $json_text = do { local $/; <$fh> };
        close $fh;
        my $state = decode_json($json_text);
        my @last_runs = keys %{$state->{last_sync}//{}};
        if (@last_runs) {
            @last_runs = sort @last_runs;
            print "📅 Letzter automatischer Sync: $last_runs[-1]\n";
        }
    }
    
    print "=" x 80 . "\n";
}

main() if __FILE__ eq $0;

File: javascript/sync_status.js
#!/usr/bin/env node
// sync_status.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:skills/sync-utils/scripts/sync_status.py
// auch in: OpenClaw@gateway2:skills/sync-utils/scripts/sync_status.py
// Erzeugt: 2026-08-20 durch ABSTRACTIONS_MANAGER.py

/**
 * Sync Status - Zeigt Status aller Skills
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

// Konstanten
const CLAWHUB_DIR = "/home/openclaw/.openclaw/workspace/skills";
const GIT_DIR = "/home/openclaw/.openclaw/workspace/git/skills";
const STATE_FILE = "/home/openclaw/.openclaw/workspace/db/sync_state.json";

/**
 * Prüft Status eines Skills
 */
function checkSkillStatus(skillName) {
    const clawhubPath = path.join(CLAWHUB_DIR, skillName);
    const gitPath = path.join(GIT_DIR, skillName);
    
    const status = {
        name: skillName,
        in_clawhub: fs.existsSync(clawhubPath),
        in_git: fs.existsSync(gitPath),
        has_git_repo: fs.existsSync(path.join(gitPath, ".git")),
        status: "unknown",
        last_modified: {}
    };
    
    // Status bestimmen
    if (status.in_clawhub && !status.in_git) {
        status.status = "only_clawhub";
    } else if (status.in_git && !status.in_clawhub) {
        status.status = "only_git";
    } else if (status.in_clawhub && status.in_git) {
        // Timestamps vergleichen
        try {
            let clawhubFiles = [];
            function walkDir(dir) {
                const files = fs.readdirSync(dir);
                for (const file of files) {
                    const filePath = path.join(dir, file);
                    const stat = fs.statSync(filePath);
                    if (stat.isDirectory()) {
                        walkDir(filePath);
                    } else {
                        clawhubFiles.push({ path: filePath, mtime: stat.mtime.getTime() });
                    }
                }
            }
            
            walkDir(clawhubPath);
            const clawhubMtimes = clawhubFiles.map(f => f.mtime);
            const clawhubMtime = Math.max(...clawhubMtimes);
            
            let gitFiles = [];
            function walkGitDir(dir) {
                const files = fs.readdirSync(dir);
                for (const file of files) {
                    if (file === '.git') continue;
                    const filePath = path.join(dir, file);
                    const stat = fs.statSync(filePath);
                    if (stat.isDirectory()) {
                        walkGitDir(filePath);
                    } else {
                        gitFiles.push({ path: filePath, mtime: stat.mtime.getTime() });
                    }
                }
            }
            
            walkGitDir(gitPath);
            const gitMtimes = gitFiles.map(f => f.mtime);
            const gitMtime = Math.max(...gitMtimes);
            
            const clawhubDate = new Date(clawhubMtime);
            const gitDate = new Date(gitMtime);
            
            status.last_modified.clawhub = clawhubDate.getFullYear() + '-' +
                String(clawhubDate.getMonth() + 1).padStart(2, '0') + '-' +
                String(clawhubDate.getDate()).padStart(2, '0') + ' ' +
                String(clawhubDate.getHours()).padStart(2, '0') + ':' +
                String(clawhubDate.getMinutes()).padStart(2, '0') + ':' +
                String(clawhubDate.getSeconds()).padStart(2, '0');
                
            status.last_modified.git = gitDate.getFullYear() + '-' +
                String(gitDate.getMonth() + 1).padStart(2, '0') + '-' +
                String(gitDate.getDate()).padStart(2, '0') + ' ' +
                String(gitDate.getHours()).padStart(2, '0') + ':' +
                String(gitDate.getMinutes()).padStart(2, '0') + ':' +
                String(gitDate.getSeconds()).padStart(2, '0');
            
            if (Math.abs(clawhubMtime - gitMtime) < 60000) {
                status.status = "synced";
            } else if (clawhubMtime > gitMtime) {
                status.status = "clawhub_newer";
            } else {
                status.status = "git_newer";
            }
        } catch (e) {
            status.status = "error";
        }
    }
    
    return status;
}

/**
 * Hauptfunktion
 */
function main() {
    console.log("=" .repeat(80));
    console.log("ClawHub ↔ Git Sync Status");
    console.log("=" .repeat(80));
    
    const now = new Date();
    console.log(`Zeitpunkt: ${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, '0')}-${String(now.getDate()).padStart(2, '0')} ${String(now.getHours()).padStart(2, '0')}:${String(now.getMinutes()).padStart(2, '0')}:${String(now.getSeconds()).padStart(2, '0')}`);
    console.log();
    
    // Alle Skills finden
    let allSkills = new Set();
    
    if (fs.existsSync(CLAWHUB_DIR)) {
        const clawhubDirs = fs.readdirSync(CLAWHUB_DIR);
        for (const dir of clawhubDirs) {
            const fullPath = path.join(CLAWHUB_DIR, dir);
            const stat = fs.statSync(fullPath);
            if (stat.isDirectory() && !dir.startsWith('.')) {
                allSkills.add(dir);
            }
        }
    }
    
    if (fs.existsSync(GIT_DIR)) {
        const gitDirs = fs.readdirSync(GIT_DIR);
        for (const dir of gitDirs) {
            const fullPath = path.join(GIT_DIR, dir);
            const stat = fs.statSync(fullPath);
            if (stat.isDirectory() && !dir.startsWith('.')) {
                allSkills.add(dir);
            }
        }
    }
    
    // Status-Kategorien
    const categories = {
        "synced": [],
        "clawhub_newer": [],
        "git_newer": [],
        "only_clawhub": [],
        "only_git": [],
        "error": []
    };
    
    // Status für jeden Skill prüfen
    for (const skill of Array.from(allSkills).sort()) {
        const status = checkSkillStatus(skill);
        categories[status.status].push(status);
    }
    
    // Ausgabe
    console.log(`📊 Gesamt: ${allSkills.size} Skills\n`);
    
    // Synchronisiert
    if (categories.synced.length > 0) {
        console.log(`✅ Synchronisiert (${categories.synced.length})`);
        for (const s of categories.synced) {
            console.log(`   - ${s.name}`);
        }
        console.log();
    }
    
    // ClawHub neuer
    if (categories.clawhub_newer.length > 0) {
        console.log(`🔄 ClawHub neuer (${categories.clawhub_newer.length})`);
        for (const s of categories.clawhub_newer) {
            console.log(`   - ${s.name} (ClawHub: ${s.last_modified.clawhub})`);
        }
        console.log();
    }
    
    // Git neuer
    if (categories.git_newer.length > 0) {
        console.log(`🔄 Git neuer (${categories.git_newer.length})`);
        for (const s of categories.git_newer) {
            console.log(`   - ${s.name} (Git: ${s.last_modified.git})`);
        }
        console.log();
    }
    
    // Nur in ClawHub
    if (categories.only_clawhub.length > 0) {
        console.log(`📦 Nur in ClawHub (${categories.only_clawhub.length})`);
        for (const s of categories.only_clawhub) {
            console.log(`   - ${s.name}`);
        }
        console.log();
    }
    
    // Nur in Git
    if (categories.only_git.length > 0) {
        console.log(`📁 Nur in Git (${categories.only_git.length})`);
        for (const s of categories.only_git) {
            console.log(`   - ${s.name}`);
        }
        console.log();
    }
    
    // Fehler
    if (categories.error.length > 0) {
        console.log(`❌ Fehler (${categories.error.length})`);
        for (const s of categories.error) {
            console.log(`   - ${s.name}`);
        }
        console.log();
    }
    
    // State-File Info
    if (fs.existsSync(STATE_FILE)) {
        const stateContent = fs.readFileSync(STATE_FILE, 'utf8');
        const state = JSON.parse(stateContent);
        const lastRuns = Object.keys(state.last_sync || {});
        if (lastRuns.length > 0) {
            console.log(`📅 Letzter automatischer Sync: ${lastRuns[lastRuns.length - 1]}`);
        }
    }
    
    console.log("=" .repeat(80));
}

main();
